File: backend/src/main/java/project/backend/domain/chat/chatmessage/entity/ChatMessage.java
package project.backend.domain.chat.chatmessage.entity;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Lob;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToOne;
import java.time.LocalDateTime;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import project.backend.domain.chat.chatroom.entity.ChatRoom;
import project.backend.domain.imagefile.ImageFile;
import project.backend.domain.member.entity.Member;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class ChatMessage {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "message_id")
	private Long id;

//	@ManyToOne
//	@JoinColumn(name = "sender_id")
//	private ChatParticipant sender;

	@ManyToOne
	@JoinColumn(name = "member_id")
	private Member sender;

	@ManyToOne
	@JoinColumn(name = "room_id")
	private ChatRoom chatRoom;

	@Lob
	@Column(columnDefinition = "TEXT")
	private String content;

	private LocalDateTime sendAt;

	@Enumerated(EnumType.STRING)
	private MessageType type = MessageType.TEXT;

	private String codeLanguage; //추가, 문법마다 다르게 하이라이팅을 하기 위함

	@OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
	@JoinColumn(name = "chat_image_id")
	private ImageFile chatImage;

	private boolean isEdited = false;
	private boolean isDeleted = false;

	@Builder
	public ChatMessage(Member sender, ChatRoom chatRoom, String content,
		LocalDateTime sendAt,
		MessageType type, String codeLanguage, ImageFile chatImage) {
		this.sender = sender;
		this.chatRoom = chatRoom;
		this.content = content;
		this.sendAt = sendAt;
		this.type = type;
		this.codeLanguage = codeLanguage;
		this.chatImage = chatImage;
	}

	public void updateContent(String newContent) {
		if (newContent != null) {
			this.content = newContent;
			isEdited = true;
		}
	}

	public void updateLanguage(String language) {
		if (language != null) {
			this.codeLanguage = language;
			isEdited = true;
		}
	}

	public void delete() {
		this.isDeleted = true;
	}
}

File: backend/src/main/java/project/backend/domain/chat/chatmessage/entity/ChatMessage.kt
package project.backend.domain.chat.chatmessage.entity

import jakarta.persistence.*
import project.backend.domain.chat.chatroom.entity.ChatRoom
import project.backend.domain.imagefile.ImageFile
import project.backend.domain.member.entity.Member
import java.time.LocalDateTime

@Entity
@Table(name = "chat_message")
class ChatMessage(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "message_id")
    val id: Long = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    val sender: Member,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "room_id")
    val chatRoom: ChatRoom,

    @Lob
    @Column(columnDefinition = "TEXT")
    var content: String? = null,

    val sendAt: LocalDateTime,

    @Enumerated(EnumType.STRING)
    val type: MessageType,

    var codeLanguage: String? = null,

    @OneToOne(cascade = [CascadeType.ALL], orphanRemoval = true)
    @JoinColumn(name = "chat_image_id")
    val chatImage: ImageFile? = null,

    @Enumerated(EnumType.STRING)
    var status: MessageStatus = MessageStatus.NO_CHANGE
) {
    protected constructor() : this(
        sender = Member(),
        chatRoom = ChatRoom(),
        sendAt = LocalDateTime.now(),
        type = MessageType.TEXT
    )

    fun updateContent(newContent: String?) {
        newContent?.let {
            content = it
            status = MessageStatus.EDITED
        }
    }

    fun updateLanguage(language: String?) {
        language?.let {
            codeLanguage = it
            status = MessageStatus.EDITED
        }
    }

    fun delete() {
        status = MessageStatus.DELETED
    }
}